### PR TITLE
AGENT-1028: Day2 - Save and manage authentication tokens as cluster secrets

### PR DESCRIPTION
**Pull request by pawanpinjarkar**
- Store the three authentication tokens (userAuth, agentAuth, watcherAuth) as a secret in the cluster when creating the node ISO.
- Automatically regenerate expired tokens and refresh the asset store to maintain valid authentication credentials in the cluster secret.

### CODE REVIEW

File: pkg/asset/agent/gencrypto/authconfig.go [hunk 1]
```diff
@@ -303,8 +312,8 @@ func (a *AuthConfig) refreshAuthTokenSecret(k8sclientset kubernetes.Interface, r
 	return nil
 }
 
-// GetAuthTokenFromCluster returns a token string stored as the secret from the cluster.
-func GetAuthTokenFromCluster(ctx context.Context, kubeconfigPath string) (string, error) {
+// GetWatcherAuthTokenFromCluster returns a watcherAuth token string stored as the secret from the cluster.
+func GetWatcherAuthTokenFromCluster(ctx context.Context, kubeconfigPath string) (string, error) {
```
Comment (rwsu) [inline pkg/asset/agent/gencrypto/authconfig.go:316]: I see there is a unit test for the install workflow. I suggest creating unit tests for the addnodes workflow. One of the unit tests should verify if the tokens are generated if they were previously generated and expired.

File: pkg/asset/agent/gencrypto/authconfig.go [hunk 2]
```diff
@@ -314,19 +323,28 @@ func GetAuthTokenFromCluster(ctx context.Context, kubeconfigPath string) (string
...
 		return "", err
 	}
-	return authToken, err
+	return watcherAuthToken, err
 }
 
-func extractAuthTokenFromSecret(secret *corev1.Secret) (string, error) {
-	existingWatcherAuthToken, exists := secret.Data[authTokenSecretDataKey]
+func extractAuthTokensFromSecret(secret *corev1.Secret) (string, string, string, error) {
+	existingAgentAuthToken, exists := secret.Data[userAuthKey]
```
Comment (andfasano) [inline pkg/asset/agent/gencrypto/authconfig.go:334]: Shouldn't be 
```suggestion
	existingAgentAuthToken, exists := secret.Data[agentAuthKey]
```

File: pkg/asset/agent/gencrypto/authconfig.go [hunk 3]
```diff
@@ -314,19 +323,28 @@ func GetAuthTokenFromCluster(ctx context.Context, kubeconfigPath string) (string
...
+	}
+
+	existingUserAuthToken, exists := secret.Data[userAuthKey]
+	if !exists || len(existingUserAuthToken) == 0 {
+		return "", "", "", fmt.Errorf("auth token secret %s/%s does not contain the key %s or is empty", authTokenSecretNamespace, authTokenSecretName, userAuthKey)
+	}
+	existingWatcherAuthToken, exists := secret.Data[watcherAuthKey]
 	if !exists || len(existingWatcherAuthToken) == 0 {
-		return "", fmt.Errorf("auth token secret %s/%s does not contain the key %s or is empty", authTokenSecretNamespace, authTokenSecretName, authTokenSecretDataKey)
+		return "", "", "", fmt.Errorf("auth token secret %s/%s does not contain the key %s or is empty", authTokenSecretNamespace, authTokenSecretName, watcherAuthKey)
```
Comment (andfasano) [inline pkg/asset/agent/gencrypto/authconfig.go:345]: Note about the migration (4.17 -> 4.18): when executed the first time against a 4.17 cluster, the `createOrUpdateAuthTokenSecret` is going to fail due the missing data keys: ie, the secret is refreshed only when it does not exist or contains all the keys. I think it should be refreshed also when it exists and contains just the old key (meaning it's a 4.17 secret).

Comment (pawanpinjarkar) [reply pkg/asset/agent/gencrypto/authconfig.go:345]: nice finding 👌

File: pkg/asset/agent/gencrypto/authconfig.go (hunk 2 above)
Comment (pawanpinjarkar) [reply pkg/asset/agent/gencrypto/authconfig.go:334]: yes, missed it

File: pkg/asset/agent/gencrypto/authconfig.go (hunk 1 above)
Comment (pawanpinjarkar) [reply pkg/asset/agent/gencrypto/authconfig.go:313]: I would like to add the new unit tests via seperate PR

**Comment by pawanpinjarkar** [comment]
hold cancel

**Pull request merged**